### ISSUE TITLE
Installing from live ISO onto EFI system

### ISSUE THREAD

**Issue by ShaunTheSalvo** [shadichy/systemback-archlinux#7]
Hey there,

I just downloaded and compiled latest source today, and created a new live ISO. Attempted to install on freshly GPT-partitioned HDD, but as yet unable to boot the system from the HDD.

Will get error logs etc for you ASAP, but I do have a question before I go any further - sorry to bother you.

Are there any specific packages I should have installed on my system (things like efibootmgr, shim-signed, etc.) prior to creating a live ISO, for SB to successfully create the /boot/efi partition?

During installations from a live ISO onto an EFI/GPT drive, I have seen the error message `EFI variables are not supported on this system`.

Apologies in advance, as I have limited experience with EFI.

**Comment by shadichy** [comment]
Sorry, but did you boot the installer in EFI mode, not BIOS/CSM mode?

**Comment by ShaunTheSalvo** [comment]
> Sorry, but did you boot the installer in EFI mode, not BIOS/CSM mode?

Yes.

At present, after installing in EFI, to make the system boot, I'm using the Debian mini.iso installer, and using the "Reinstall GRUB bootloader". Once this is done, the system will boot.

**Comment by shadichy** [comment]
> During installations from a live ISO onto an EFI/GPT drive, I have seen the error message `EFI variables are not supported on this system`.

Did you check for "/sys/firmware/efi"?

**Comment by shadichy** [comment]
> > Sorry, but did you boot the installer in EFI mode, not BIOS/CSM mode?
> 
> Yes.
> 
> At present, after installing in EFI, to make the system boot, I'm using the Debian mini.iso installer, and using the "Reinstall GRUB bootloader". Once this is done, the system will boot.

If you do boot the installer in EFI mode, then the folder `/sys/firmware/efi` will be available
![image](https://user-images.githubusercontent.com/60534636/185562887-bae1e9dc-6be5-4862-bfd3-63e083c1cd56.png)

**Comment by ShaunTheSalvo** [comment]
Hi, hopefully this will provide the info you need.

Apologies for the brevity of my reply: I do see you've updated the source in the last hour or so, so I will build and test again later tonight (have to go out for a bit right now).

Output of `ls /sys/firmware/efi -R`:
[ls-sys-firmware-efi.txt](https://github.com/shadichy/systemback-archlinux/files/9380266/ls-sys-firmware-efi.txt)

Error message from SB on completion of install under EFI:
![SB-error_on_install_completion](https://user-images.githubusercontent.com/110704527/185572907-0898c9f4-983b-4174-9e20-a60e909fb1e5.png)

Log from terminal (konsole) during SB install under EFI:
[SB-terminal-log-during-install.txt](https://github.com/shadichy/systemback-archlinux/files/9380267/SB-terminal-log-during-install.txt)

`/sys/firmware/efi` from Midnight Commander (mc):
![sys-firmware-efi dir](https://user-images.githubusercontent.com/110704527/185572916-c26a4d99-5ec5-4a19-9168-e4d5f0f2b8ec.png)

**Comment by shadichy** [comment]
Sorry about that, maybe /boot/efi is not created during install

**Comment by ShaunTheSalvo** [comment]
> Sorry about that, maybe /boot/efi is not created during install

No problem. I'll rebuild again tomorrow, and see if /boot/efi is getting created, and if so, what's in it. Not feeling 100% tonight, so I think an early night might be in order.

As an aside --- to simplify/automate the testing process, I've made a little script to uninstall SB, download fresh copy of source, and build/install. Works like a charm. :)

**Comment by shadichy** [comment]
> During installations from a live ISO onto an EFI/GPT drive, I have seen the error message `EFI variables are not supported on this system`.

This is because `/sys` is not properly mounted on chroot during grub installation. It causes `efivarfs` not to be mounted. The problem is fixed in the latest commit.

**Comment by ShaunTheSalvo** [comment]
> > During installations from a live ISO onto an EFI/GPT drive, I have seen the error message `EFI variables are not supported on this system`.
> 
> This is because `/sys` is not properly mounted on chroot during grub installation. It causes `efivarfs` not to be mounted. The problem is fixed in the latest commit.

Nice! - What the heck, I'll download and test now. :)

**Comment by ShaunTheSalvo** [comment]
Good morning,

I made a new ISO using the updated SB, and reinstalled. The new installation is recognised in the BIOS (I can see the drive it's been installed to in my BIO boot menu), but I can't get a GRUB menu to come up. I still need to use the Debian mini.iso to correct the GRUB installation.

I tried using different sizes of EFI partition, did not seem to make any difference. Will test more later today.

**Comment by shadichy** [comment]
> Good morning,
> 
> I made a new ISO using the updated SB, and reinstalled. The new installation is recognised in the BIOS (I can see the drive it's been installed to in my BIO boot menu), but I can't get a GRUB menu to come up. I still need to use the Debian mini.iso to correct the GRUB installation.
> 
> I tried using different sizes of EFI partition, did not seem to make any difference. Will test more later today.

Do you have screenshots or any pictures? Or systemback output?

**Comment by ShaunTheSalvo** [comment]
Here's the grub.cfg from an install I did to my external hdd last night. Hope this helps.

I'm sorry I can't get any screenshots, there's really nothing to see except a blank screen :). I'll do another install later, and try to grab logs from SB. Apologies for the delay, I'm still feel pretty unwell today. May not be able to get stuff back to you until tomorrow

[grub-cfg.txt](https://github.com/shadichy/systemback-archlinux/files/9386218/grub-cfg.txt)
.

**Comment by ShaunTheSalvo** [comment]
Hey - just have an update for you, and a workaround that I have found works for me. :) 

During the install process, SB seems to be trying to install GRUB for i386-platform, instead of x86_64. Snippet from terminal log of install:

```
Installing for i386-efi platform.
grub-install: error: failed to get canonical path of `/boot/efi'.
Generating grub configuration file ...
Found linux image: /boot/vmlinuz-linux
Found initrd image: /boot/initramfs-linux.img
Found fallback initrd image(s) in /boot:  initramfs-linux-fallback.img
Warning: os-prober will not be executed to detect other bootable partitions.
Systems on them will not be added to the GRUB boot configuration.
Check GRUB_DISABLE_OS_PROBER documentation entry.
done
```

Also, SB will fail if I try to format the /boot/efi partition during installation.

What I have found works is this little workaround method (I'm typing this from my system installed just now using this method):

- From SB installer, create new partition table, including a vfat partition to use for /boot/efi. NB: I found I can get away with a /boot/efi partition as small as 16mb. Of course, also create a / and /home partition (and swap if required). Set all partitions to be formatted.
My partition table for my (puny) 120GB SSD drive is:
/dev/sda1 - 16mb FAT16 - /boot/efi
/dev/sda2 - 10gb EXT4 - / (system)
/dev/sda3 - 115gb (or whatever) EXT4 - /home

- Attempt to proceed with install. SB will fail, reporting that it cannot format /dev/sda1.
- Go back to the partitioning screen, reset the mount points (click the circle-arrow button), and re-enter the partition table as above, EXCEPT I set /dev/sda1 (/boot/efi) to NOT be formatted.
- Attempt the install again, and it will work. Exit SB **but don't reboot yet** (EFI will pick up the installation, but will not boot, as GRUB is not properly installed).
- Open a terminal, and run:
```
sudo mount /dev/sda2 /mnt
sudo mkdir -p /mnt/boot/efi
sudo mount /dev/sda1 /mnt/boot/efi
sudo arch-chroot /mnt
(from within arch-chroot) grub-install /dev/sda
exit (to leave arch-chroot)
sudo reboot
```

The system will usually boot successfully using this method. I've tried repeating this a few times, but occasionally it will fail (GRUB menu will not appear on reboot). I have no idea why.

Hoping this helps. :) Have a great night.

**Comment by shadichy** [comment]
Hi, can I have full installation log please?

> `grub-install: error: failed to get canonical path of `/boot/efi'.`

Also, at that point, is "/boot/efi" mounted, or does the dedicated block device contains `fat` file system?

**Comment by ShaunTheSalvo** [comment]
> Hi, can I have full installation log please?

Certainly! Here 'tis...
[SB-terminal-log-during-install.txt](https://github.com/shadichy/systemback-archlinux/files/9387116/SB-terminal-log-during-install.txt)

> > `grub-install: error: failed to get canonical path of `/boot/efi'.`
> 
> Also, at that point, is "/boot/efi" mounted, or does the dedicated block device contains `fat` file system?

To be honest, I don't really know if it's mounted at that point. I don't think it is. I'm not really sure how to find out, as that point of the process runs through pretty quickly.

I've run into that same error now and then when attempting to do a grub-install manually on an EFI partition layout (not through Systemback), until I figured out the process I mentioned in my last post just this evening.

The block device does seem to have a 'fat' file system. Despite the first SB install attempt failing with an error, it seems to be formatting the partition anyway, as the second install attempt (without explicitly requesting the EFI partition be formatted) runs with no problem, and the manual 'grub-install' does not produce any errors.

I can try installing again tomorrow sometime, and see if I can find out if the /boot/efi is mounted at the time.

**Comment by shadichy** [comment]
```

 An error occurred while mounting the following partition/image:

  /dev/sdd1

inInstalling for i386-efi platform.
grub-install: error: failed to get canonical path of `/boot/efi'.
```
Now I know why. `/boot/efi` seems not to be mounted during grub install. Wasn't it formatted `fat` before install? According to your log, the answer is maybe not. This is original `systemback` related problem. I'll try to fix it and request upstream push for my patch (if i can)

**Comment by ShaunTheSalvo** [comment]
> ```
> 
>  An error occurred while mounting the following partition/image:
> 
>   /dev/sdd1
> 
> inInstalling for i386-efi platform.
> grub-install: error: failed to get canonical path of `/boot/efi'.
> ```
> 
> Wasn't it formatted `fat` before install? According to your log, the answer is maybe not. 

Now that you mention it, I recall the original Debian Systemback had the same problem when trying to install to EFI-based system. Sorry I didn't send the entire log in the first place.

To be honest, I can't think of anything else that needs fixing/attention. If there's anything specific you'd like me to test, let me know.

I have also tried partitioning and formatting the drive using `gparted` before running SB, including formatting the EFI partition - and then running SB to do the install. The same error occurs - SB cannot format the EFI partition (will fail with error message) if I ask it to format the EFI partition. But it will NOT fail if I just mount the EFI partition (already formatted as FAT) without formatting. However - again, the resultant installed system will not boot unless I manually install GRUB, either through terminal or via the Debian mini.iso. Much the same result with either Debian or Arch version of SB.

Best of luck - if there's any further info etc. you need from me, let me know. Have a great day.

**Comment by shadichy** [comment]
>  The same error occurs - SB cannot format the EFI partition (will fail with error message) if I ask it to format the EFI partition.

Can you show me the log please?

**Comment by ShaunTheSalvo** [comment]
> > The same error occurs - SB cannot format the EFI partition (will fail with error message) if I ask it to format the EFI partition.
> 
> Can you show me the log please?

Yep, for sure (BTW sorry for delay in reply). Here it is:

[format-efi-partition-error.txt](https://github.com/shadichy/systemback-archlinux/files/9388819/format-efi-partition-error.txt)

My guess is the error is in this line:
`mkfs.vfat: Labels with characters *?.,;:/\|+=<>[]" are not allowed'

This is the error pop-up it shows:
![format-efi-partition-errormessage](https://user-images.githubusercontent.com/110704527/185785553-55e1fbc8-3cee-47cf-ba8a-cd58fedd39ef.png)

It doesn't give me any say in what the partition label should be.

Nonetheless, it seems to be formatting the partition anyway. As I mentioned yesterday, if I then go back to the partitioning screen, reset the mountpoints by clicking the round green arrow button, and reassign the mountpoints but select the EFI partition to _not_ be formatted, the error does not occur, and the installation will proceed.

**Comment by shadichy** [comment]
Can you rebuild from source again and retry?

**Comment by ShaunTheSalvo** [comment]
> Can you rebuild from source again and retry?

Wow! Very nice indeed! The whole creation and restoration of images seems to be working pretty much perfectly. 

Made a new live ISO, and can install it with absolutely no problems onto EFI - as long as I use a minimum 32mb EFI partition. Install runs through without any problems, and the /boot/efi files load up without any problems. The installed system booted first time.

The only little thing was a minor bug seems to have cropped up during the build/install of SB itself. After building from source, the actual install failed, as files `systemback.excludes` and `systemback.includes` already existed in `/etc/systemback`. It seems some part of the install process (not the build process) is creating these files before SB itself is installed. Builds prior to today's would compile and install with no problem.

I tried deleting `/etc/systemback` altogether before reinstalling the binary packages (`sudo pacman -U systemback-1.8.9-3-any.pkg.tar.zst'), but the installer still failed.

What I managed to do as a workaround was to delete the directory (`sudo rm -rf /etc/systemback`) at the point that pacman asked me for my password - then provide my password, and everything installed just fine.

Once installed, everything is fine. In fact, I'd say SB is now running even better than the original Debian version (I was using BlueWhaleRobot's build for months on Debian, with no issues).

I did have a little trouble last night with trying to restore from a restore point. I'm trying again now, and if the problem still exists, I'll post it in a separate thread.

Once again, I really appreciate your hard work on this project - you've done an excellent job, and IMHO having a tool like Systemback finally available for Arch is a massive step forward for the distro in itself. Given Systemback's ridiculously easy interface for installing a created image, installing Arch has never been so fast and easy.

By the way - my friend Gene, who has recently started using Arch and has been helping me with some testing, sends his greetings, and his appreciation for making this tool available. Between us, we've been able to test your version of SB on a few different PC's with varying configurations (EFI, legacy bios, different amounts of RAM etc.).

I'll mark this thread as closed. Thanks again.
